File: resources/shared.go
package resources

import "time"

// Attachment represents a file attachment in a message
type Attachment struct {
	Type        string          `json:"type"` // unknown, img, video, audio
	Duration    *int            `json:"duration,omitempty"`
	FileName    *string         `json:"fileName,omitempty"`
	FileSize    *int64          `json:"fileSize,omitempty"`
	IsGif       *bool           `json:"isGif,omitempty"`
	IsSticker   *bool           `json:"isSticker,omitempty"`
	IsVoiceNote *bool           `json:"isVoiceNote,omitempty"`
	MimeType    *string         `json:"mimeType,omitempty"`
	PosterImg   *string         `json:"posterImg,omitempty"`
	Size        *AttachmentSize `json:"size,omitempty"`
	SrcURL      *string         `json:"srcURL,omitempty"`
}

// AttachmentSize represents pixel dimensions of an attachment
type AttachmentSize struct {
	Height *int `json:"height,omitempty"`
	Width  *int `json:"width,omitempty"`
}

// BaseResponse represents a basic API response
type BaseResponse struct {
	Success bool    `json:"success"`
	Error   *string `json:"error,omitempty"`
}

// Error represents an API error response
type ErrorResponse struct {
	Error   string            `json:"error"`
	Code    *string           `json:"code,omitempty"`
	Details map[string]string `json:"details,omitempty"`
}

// Message represents a chat message
type Message struct {
	ID          string       `json:"id"`
	AccountID   string       `json:"accountID"`
	ChatID      string       `json:"chatID"`
	MessageID   string       `json:"messageID"`
	SenderID    string       `json:"senderID"`
	SortKey     interface{}  `json:"sortKey"` // string or number
	Timestamp   time.Time    `json:"timestamp"`
	Attachments []Attachment `json:"attachments,omitempty"`
	IsSender    *bool        `json:"isSender,omitempty"`
	IsUnread    *bool        `json:"isUnread,omitempty"`
	Reactions   []Reaction   `json:"reactions,omitempty"`
	SenderName  *string      `json:"senderName,omitempty"`
	Text        *string      `json:"text,omitempty"`
}

// Reaction represents a message reaction
type Reaction struct {
	ID            string  `json:"id"`
	ParticipantID string  `json:"participantID"`
	ReactionKey   string  `json:"reactionKey"`
	Emoji         *bool   `json:"emoji,omitempty"`
	ImgURL        *string `json:"imgURL,omitempty"`
}

// User represents a person on or reachable through Beeper
type User struct {
	ID            string  `json:"id"`
	CannotMessage *bool   `json:"cannotMessage,omitempty"`
	Email         *string `json:"email,omitempty"`
	FullName      *string `json:"fullName,omitempty"`
	ImgURL        *string `json:"imgURL,omitempty"`
	IsSelf        *bool   `json:"isSelf,omitempty"`
	PhoneNumber   *string `json:"phoneNumber,omitempty"`
	Username      *string `json:"username,omitempty"`
}

// Cursor represents a paginated response
type Cursor[T any] struct {
	Items      []T             `json:"items"`
	Pagination *PaginationInfo `json:"pagination,omitempty"`
}

// PaginationInfo contains pagination metadata
type PaginationInfo struct {
	Cursor    *string `json:"cursor,omitempty"`
	Limit     *int    `json:"limit,omitempty"`
	Direction *string `json:"direction,omitempty"`
	HasMore   bool    `json:"has_more"`
}

// MessagesCursor is a type alias for message pagination
type MessagesCursor = Cursor[Message]

File: beeper-rust-sdk/src/resources/shared.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Attachment represents a file attachment in a message
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Attachment {
    /// Type of attachment (unknown, img, video, audio)
    #[serde(rename = "type")]
    pub attachment_type: String,
    pub duration: Option<i32>,
    pub file_name: Option<String>,
    pub file_size: Option<i64>,
    pub is_gif: Option<bool>,
    pub is_sticker: Option<bool>,
    pub is_voice_note: Option<bool>,
    pub mime_type: Option<String>,
    pub poster_img: Option<String>,
    pub size: Option<AttachmentSize>,
    pub src_url: Option<String>,
}

/// AttachmentSize represents pixel dimensions of an attachment
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AttachmentSize {
    pub height: Option<i32>,
    pub width: Option<i32>,
}

/// BaseResponse represents a basic API response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BaseResponse {
    pub success: bool,
    pub error: Option<String>,
}

/// Error represents an API error response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorResponse {
    pub error: String,
    pub code: Option<String>,
    pub details: Option<HashMap<String, String>>,
}

/// Message represents a chat message
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub id: String,
    #[serde(rename = "accountID")]
    pub account_id: String,
    #[serde(rename = "chatID")]
    pub chat_id: String,
    #[serde(rename = "messageID")]
    pub message_id: String,
    #[serde(rename = "senderID")]
    pub sender_id: String,
    #[serde(rename = "sortKey")]
    pub sort_key: serde_json::Value, // Can be string or number
    pub timestamp: DateTime<Utc>,
    pub attachments: Option<Vec<Attachment>>,
    #[serde(rename = "isSender")]
    pub is_sender: Option<bool>,
    #[serde(rename = "isUnread")]
    pub is_unread: Option<bool>,
    pub reactions: Option<Vec<Reaction>>,
    #[serde(rename = "senderName")]
    pub sender_name: Option<String>,
    pub text: Option<String>,
}

/// Reaction represents a message reaction
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Reaction {
    pub id: String,
    pub participant_id: String,
    pub reaction_key: String,
    pub emoji: Option<bool>,
    pub img_url: Option<String>,
}

/// User represents a person on or reachable through Beeper
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct User {
    pub id: String,
    #[serde(rename = "cannotMessage")]
    pub cannot_message: Option<bool>,
    pub email: Option<String>,
    #[serde(rename = "fullName")]
    pub full_name: Option<String>,
    #[serde(rename = "imgURL")]
    pub img_url: Option<String>,
    #[serde(rename = "isSelf")]
    pub is_self: Option<bool>,
    #[serde(rename = "phoneNumber")]
    pub phone_number: Option<String>,
    pub username: Option<String>,
}

/// Cursor represents a paginated response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Cursor<T> {
    pub items: Vec<T>,
    pub pagination: Option<PaginationInfo>,
}

/// PaginationInfo contains pagination metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PaginationInfo {
    pub cursor: Option<String>,
    pub limit: Option<i32>,
    pub direction: Option<String>,
    pub has_more: bool,
}

/// MessagesCursor is a type alias for message pagination
pub type MessagesCursor = Cursor<Message>;

/// ChatsCursor represents paginated chat results
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatsCursor {
    pub items: Vec<Chat>,
    pub pagination: Option<PaginationInfo>,
}

/// Chat represents a chat/conversation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Chat {
    pub id: String,
    #[serde(rename = "accountID")]
    pub account_id: String,
    pub network: String,
    pub title: String,
    /// Type of chat (single, group)
    #[serde(rename = "type")]
    pub chat_type: String,
    #[serde(rename = "unreadCount")]
    pub unread_count: i32,
    pub participants: ChatParticipants,
    #[serde(rename = "isArchived")]
    pub is_archived: Option<bool>,
    #[serde(rename = "isMuted")]
    pub is_muted: Option<bool>,
    #[serde(rename = "isPinned")]
    pub is_pinned: Option<bool>,
    #[serde(rename = "lastActivity")]
    pub last_activity: Option<String>,
    #[serde(rename = "lastReadMessageSortKey")]
    pub last_read_message_sort_key: Option<serde_json::Value>, // Can be string or number
    #[serde(rename = "localChatID")]
    pub local_chat_id: Option<String>,
}

/// ChatParticipants represents chat participants information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatParticipants {
    #[serde(rename = "hasMore")]
    pub has_more: bool,
    pub items: Vec<User>,
    pub total: i32,
}

/// Account represents a chat account added to Beeper
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Account {
    #[serde(rename = "accountID")]
    pub account_id: String,
    pub network: String,
    pub user: User,
}